diff --git a/src/backend/tcop/postgres.c b/src/backend/tcop/postgres.c
index 470b734e9e..0798dff410 100644
--- a/src/backend/tcop/postgres.c
+++ b/src/backend/tcop/postgres.c
@@ -73,6 +73,9 @@
 #include "utils/snapmgr.h"
 #include "utils/timeout.h"
 #include "utils/timestamp.h"
+#ifdef USE_VALGRIND
+#include <valgrind/valgrind.h>
+#endif
 
 /* ----------------
  *		global variables
@@ -986,6 +989,7 @@ exec_simple_query(const char *query_string)
 	bool		was_logged = false;
 	bool		use_implicit_block;
 	char		msec_str[32];
+	static long long unsigned vgErrorsSoFar = 0;
 
 	/*
 	 * Report query to various monitoring facilities.
@@ -1343,6 +1347,16 @@ exec_simple_query(const char *query_string)
 	TRACE_POSTGRESQL_QUERY_DONE(query_string);
 
 	debug_query_string = NULL;
+
+#ifdef USE_VALGRIND
+	if (VALGRIND_COUNT_ERRORS > vgErrorsSoFar)
+	{
+		VALGRIND_PRINTF("The query for which valgrind reported a "
+						"memory error was: %s\n",
+						query_string);
+	}
+	vgErrorsSoFar = VALGRIND_COUNT_ERRORS;
+#endif
 }
 
 /*
